tests/misc/state.py

```diff
@@ -0,0 +1,214 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+#
+# Copyright (C) Pootle contributors.
+#
+# This file is a part of the Pootle project. It is distributed under the GPL3
+# or later license. See the LICENSE file for a copy of the license and the
+# AUTHORS file for copyright and authorship information.
+
+import pytest
+
+from pootle.core.state import State, ItemState
+
+
+class DummyContext(object):
+
+    def __str__(self):
+        return "<DummyContext object>"
+
+
+@pytest.mark.django_db
+def test_state_instance():
+
+    context = DummyContext()
+    state = State(context)
+
+    assert state.context == context
+    assert state.__state__ == {}
+    assert state.prefix == "state"
+    assert state.has_changed is False
+    assert state.states == []
+    assert "x" not in state
+    assert list(state) == []
+    assert state.item_state_class == ItemState
+    assert str(state) == (
+        "<State(<DummyContext object>): Everything up-to-date>")
+
+
+@pytest.mark.django_db
+def test_state_states():
+
+    class ContextualState(State):
+
+        @property
+        def states(self):
+            return ["foo", "bar", "empty"]
+
+        def state_foo(self, **kwargs):
+            for x in [1, 2, 3]:
+                yield {str(x): x}
+
+        def state_bar(self, **kwargs):
+            for x in [4, 5, 6, 7, 8, 9]:
+                yield {str(x): x}
+
+        def state_baz(self, **kwargs):
+            yield dict(never="called")
```

Comment:
im happy to use dict(), esp in a test

there are situations when an explicit `{}` is better (ie when you need to use vars for keys), but on the whole, i prefer using the `dict` builtin